From 93dcfd741edcf99b2dfc3923d2642974d5599361 Mon Sep 17 00:00:00 2001
From: Mikhail Paulyshka <me@mixaill.net>
Date: Thu, 13 Mar 2025 01:21:59 +0300
Subject: [PATCH 4/4] x86/rdrand: hide RDRAND and RDSEED from CPUID in case of
 a malfunction

Some applications use the CPUID call instead of /proc/cpuinfo to get
CPU capabilities. In this case, they will still try to use non-functional
RDRAND/RDSEED implementations.

Disables visibility of RDRAND and RDSEED on AMD platforms.

Signed-off-by: Mikhail Paulyshka <me@mixaill.net>
---
 arch/x86/kernel/cpu/rdrand.c | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/arch/x86/kernel/cpu/rdrand.c b/arch/x86/kernel/cpu/rdrand.c
index 2184340a0d3c..5769d2cab1d6 100644
--- a/arch/x86/kernel/cpu/rdrand.c
+++ b/arch/x86/kernel/cpu/rdrand.c
@@ -43,6 +43,8 @@ void x86_init_rdrand(struct cpuinfo_x86 *c)
 
 		if (failure) {
 			clear_cpu_cap(c, X86_FEATURE_RDRAND);
+			if (c->x86_vendor == X86_VENDOR_AMD || c->x86_vendor == X86_VENDOR_HYGON)
+				msr_clear_bit(MSR_AMD64_CPUID_FN_1, 62);
 			pr_emerg("RDRAND is not reliable on this platform; disabling.\n");
 		}
 	}
@@ -62,6 +64,8 @@ void x86_init_rdrand(struct cpuinfo_x86 *c)
 
 		if (failure) {
 			clear_cpu_cap(c, X86_FEATURE_RDSEED);
+			if (c->x86_vendor == X86_VENDOR_AMD || c->x86_vendor == X86_VENDOR_HYGON)
+				msr_clear_bit(MSR_AMD64_CPUID_FN_7, 18);
 			pr_emerg("RDSEED is not reliable on this platform; disabling.\n");
 		}
 	}
-- 
2.48.1

